Review Analysis of Amazon Dataset

In [2]:
!pip install nltk

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 9.9 MB/s eta 0:00:00


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import re

from nltk.tokenize import word_tokenize
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /Users/nisha/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [28]:
nltk.download('punkt_tab')

[nltk_data] Downloading package punkt_tab to /Users/nisha/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


True

In [37]:
from nltk.corpus import stopwords
nltk.download('stopwords')

stop_words = set(stopwords.words('english'))

[nltk_data] Downloading package stopwords to /Users/nisha/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [2]:
df = pd.read_csv('amazon_review_full_csv/train.csv', delimiter=',', header=None)

In [3]:
df.head()

0                                      1  \
0  3                     more like funchuck   
1  5                              Inspiring   
2  5  The best soundtrack ever to anything.   
3  4                       Chrono Cross OST   
4  5                    Too good to be true   

                                                   2  
0  Gave this to my dad for a gag gift after direc...  
1  I hope a lot of people hear this cd. We need m...  
2  I'm reading a lot of reviews saying that this ...  
3  The music of Yasunori Misuda is without questi...  
4  Probably the greatest soundtrack in history! U...

In [4]:
df.columns = ['Rating',"Review_Title", "Review"]

In [5]:
df.shape

(3000000, 3)

In [6]:
df.head()

Rating                           Review_Title  \
0       3                     more like funchuck   
1       5                              Inspiring   
2       5  The best soundtrack ever to anything.   
3       4                       Chrono Cross OST   
4       5                    Too good to be true   

                                              Review  
0  Gave this to my dad for a gag gift after direc...  
1  I hope a lot of people hear this cd. We need m...  
2  I'm reading a lot of reviews saying that this ...  
3  The music of Yasunori Misuda is without questi...  
4  Probably the greatest soundtrack in history! U...

<h1>DATA WRANGLING</h1>

# 1. Check for missing values

In [7]:
df.isnull().mean()*100

Rating          0.000000
Review_Title    0.006267
Review          0.000000
dtype: float64

In [8]:
df.isnull().sum()

Rating            0
Review_Title    188
Review            0
dtype: int64

# Out of 3000000, there are 188 missing values , Since the missing values are in title, it may not make difference in finding the analysis, so this can be ignored.

# 2. Checking DataTyes are correct.

In [9]:
data_types = df.dtypes
print(data_types)

Rating           int64
Review_Title    object
Review          object
dtype: object


In [10]:
# converting the Columns to appropriate dataType for Review_Title and Review
df["Review_Title"] = df["Review_Title"].astype(str)
df["Review"] = df["Review"].astype(str)

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000000 entries, 0 to 2999999
Data columns (total 3 columns):
 #   Column        Dtype 
---  ------        ----- 
 0   Rating        int64 
 1   Review_Title  object
 2   Review        object
dtypes: int64(1), object(2)
memory usage: 68.7+ MB


In [12]:
df['Review_Title'] = df['Review_Title'].astype(str)
df["Review"] = df["Review"].astype(str)

In [13]:
df.dtypes

Rating           int64
Review_Title    object
Review          object
dtype: object

In [14]:
print(type(df['Review_Title'][0]))  # Should print <class 'str'>
print(type(df['Review'][0]))        # Should print <class 'str'>

<class 'str'>
<class 'str'>


In [15]:
# checking Duplicate values

duplicated = df[df.duplicated()]
print(duplicated)

Empty DataFrame
Columns: [Rating, Review_Title, Review]
Index: []


In [16]:
# Checking Duplicates based on specific column
dup1 = df[df.duplicated(subset=['Rating','Review_Title'])]
dup2 = df[df.duplicated(subset=['Rating','Review'])]

In [17]:
# There are no Duplicate Reviews

In [18]:
print(dup1)

         Rating     Review_Title  \
95            2     Not the best   
103           1  I got fooled...   
202           2    Disappointing   
297           3   Mixed feelings   
555           2     disappointed   
...         ...              ...   
2999986       3       Incomplete   
2999987       2  Not the Best...   
2999992       2   Too simplistic   
2999995       1    Don't do it!!   
2999998       3   Hard to clean!   

                                                    Review  
95       I agree, the CDs are a much needed help, since...  
103      I did not know this was the audio cd of the ga...  
202      I bought this book because I am a nonogram (al...  
297      Some beautiful scenery early on and nice roman...  
555      I think this is a case of my expectations bein...  
...                                                    ...  
2999986  The three films in this set are excellent, abs...  
2999987  I bought this CD expecting it to be the same t...  
2999992  While Mr.

In [19]:
dup2.head()

Rating                                       Review_Title  \
1072        1                                   Wouldn't stay up   
7002        2                     lesson in the depravity of man   
12273       3                                      Shows Its Age   
15487       1  Dumb adaptation that does not try to be even r...   
15531       1                                    What a letdown.   

                                                  Review  
1072   The band was far too thin and would roll down ...  
7002   As I opened this book, I thought to myself "Wo...  
12273  Unfortunately, "20,000 Leagues Under the Sea" ...  
15487  Lackluster adaptation of the vertigo comic boo...  
15531  Shameful attempt at a hybrid of horror, action...

In [20]:
# doubt - are these duplicates ?

# NLP - WRANGLING

In [21]:
# 1.Convert String to lowercase
df['Review_Title'] = df['Review_Title'].str.lower()
df['Review'] = df['Review'].str.lower()

In [22]:
df.head()

Rating                           Review_Title  \
0       3                     more like funchuck   
1       5                              inspiring   
2       5  the best soundtrack ever to anything.   
3       4                       chrono cross ost   
4       5                    too good to be true   

                                              Review  
0  gave this to my dad for a gag gift after direc...  
1  i hope a lot of people hear this cd. we need m...  
2  i'm reading a lot of reviews saying that this ...  
3  the music of yasunori misuda is without questi...  
4  probably the greatest soundtrack in history! u...

In [23]:
df_1 = df.copy()

In [24]:
# 2. Remove punctuation
def removeall_exceptStringSpace(text):
    eos_text = re.sub(r'[^\w\s\.]', '', text)
    eos_text =re.sub(r'\.','<eos>',text)
    return eos_text

df_1['Review_Title'] = df_1['Review_Title'].apply(removeall_exceptStringSpace)
df_1['Review'] = df_1['Review'].apply(removeall_exceptStringSpace)

In [25]:
df_1.head()

Rating                               Review_Title  \
0       3                         more like funchuck   
1       5                                  inspiring   
2       5  the best soundtrack ever to anything<eos>   
3       4                           chrono cross ost   
4       5                        too good to be true   

                                              Review  
0  gave this to my dad for a gag gift after direc...  
1  i hope a lot of people hear this cd<eos> we ne...  
2  i'm reading a lot of reviews saying that this ...  
3  the music of yasunori misuda is without questi...  
4  probably the greatest soundtrack in history! u...

In [29]:
# Tokenize text columns
df_1['Review_Title_Tokens'] = df_1['Review_Title'].apply(word_tokenize)
df_1['Review_Tokens'] = df_1['Review'].apply(word_tokenize)

In [30]:
df_1.head()

Rating                               Review_Title  \
0       3                         more like funchuck   
1       5                                  inspiring   
2       5  the best soundtrack ever to anything<eos>   
3       4                           chrono cross ost   
4       5                        too good to be true   

                                              Review  \
0  gave this to my dad for a gag gift after direc...   
1  i hope a lot of people hear this cd<eos> we ne...   
2  i'm reading a lot of reviews saying that this ...   
3  the music of yasunori misuda is without questi...   
4  probably the greatest soundtrack in history! u...   

                                 Review_Title_Tokens  \
0                             [more, like, funchuck]   
1                                        [inspiring]   
2  [the, best, soundtrack, ever, to, anything, <,...   
3                               [chrono, cross, ost]   
4                          [too, good, to, be, true]   

                                       Review_Tokens  
0  [gave, this, to, my, dad, for, a, gag, gift, a...  
1  [i, hope, a, lot, of, people, hear, this, cd, ...  
2  [i, 'm, reading, a, lot, of, reviews, saying, ...  
3  [the, music, of, yasunori, misuda, is, without...  
4  [probably, the, greatest, soundtrack, in, hist...

In [32]:
# Check for Text length

df_1['Review_Length'] = df_1['Review_Tokens'].apply(len)

In [34]:
df_1['Review_Title_Length'] = df_1['Review_Title_Tokens'].apply(len)

In [35]:
df_1.head()

Rating                               Review_Title  \
0       3                         more like funchuck   
1       5                                  inspiring   
2       5  the best soundtrack ever to anything<eos>   
3       4                           chrono cross ost   
4       5                        too good to be true   

                                              Review  \
0  gave this to my dad for a gag gift after direc...   
1  i hope a lot of people hear this cd<eos> we ne...   
2  i'm reading a lot of reviews saying that this ...   
3  the music of yasunori misuda is without questi...   
4  probably the greatest soundtrack in history! u...   

                                 Review_Title_Tokens  \
0                             [more, like, funchuck]   
1                                        [inspiring]   
2  [the, best, soundtrack, ever, to, anything, <,...   
3                               [chrono, cross, ost]   
4                          [too, good, to, be, true]   

                                       Review_Tokens  Review_Title_Length  \
0  [gave, this, to, my, dad, for, a, gag, gift, a...                    3   
1  [i, hope, a, lot, of, people, hear, this, cd, ...                    1   
2  [i, 'm, reading, a, lot, of, reviews, saying, ...                    9   
3  [the, music, of, yasunori, misuda, is, without...                    3   
4  [probably, the, greatest, soundtrack, in, hist...                    5   

   Review_Length  
0             24  
1             54  
2            111  
3             86  
4             81

In [45]:
df_1['Review_Title_Tokens_stop_words_removed'] = df_1['Review_Title_Tokens'].apply(lambda x: [word for word in x if word not in stop_words])
df_1['Review_Tokens_stop_words_removed'] = df_1['Review_Tokens'].apply(lambda x: [word for word in x if word not in stop_words])

In [46]:
df_1.head()

Rating                               Review_Title  \
0       3                         more like funchuck   
1       5                                  inspiring   
2       5  the best soundtrack ever to anything<eos>   
3       4                           chrono cross ost   
4       5                        too good to be true   

                                              Review  \
0  gave this to my dad for a gag gift after direc...   
1  i hope a lot of people hear this cd<eos> we ne...   
2  i'm reading a lot of reviews saying that this ...   
3  the music of yasunori misuda is without questi...   
4  probably the greatest soundtrack in history! u...   

                                 Review_Title_Tokens  \
0                             [more, like, funchuck]   
1                                        [inspiring]   
2  [the, best, soundtrack, ever, to, anything, <,...   
3                               [chrono, cross, ost]   
4                          [too, good, to, be, true]   

                                       Review_Tokens  Review_Title_Length  \
0  [gave, this, to, my, dad, for, a, gag, gift, a...                    3   
1  [i, hope, a, lot, of, people, hear, this, cd, ...                    1   
2  [i, 'm, reading, a, lot, of, reviews, saying, ...                    9   
3  [the, music, of, yasunori, misuda, is, without...                    3   
4  [probably, the, greatest, soundtrack, in, hist...                    5   

   Review_Length         Review_Title_Tokens_stop_words_removed  \
0             24                               [like, funchuck]   
1             54                                    [inspiring]   
2            111  [best, soundtrack, ever, anything, <, eos, >]   
3             86                           [chrono, cross, ost]   
4             81                                   [good, true]   

                    Review_Tokens_stop_words_removed  
0  [gave, dad, gag, gift, directing, ``, nunsense...  
1  [hope, lot, people, hear, cd, <, eos, >, need,...  
2  ['m, reading, lot, reviews, saying, best, 'gam...  
3  [music, yasunori, misuda, without, question, c...  
4  [probably, greatest, soundtrack, history, !, u...

In [47]:
df_1.to_csv('data_wrangled_reviews_amazon.csv', index=False)